core/ts/Poll.ts

```diff
@@ -423,159 +428,201 @@ export class Poll implements IPoll {
     }
 
     // The starting index must be valid
-    assert(this.currentMessageBatchIndex >= 0);
-    assert(this.currentMessageBatchIndex % batchSize === 0);
+    assert(this.currentMessageBatchIndex >= 0, "The starting index must be >= 0");
+    assert(this.currentMessageBatchIndex % batchSize === 0, "The starting index must be a multiple of the batch size");
 
+    // ensure we copy the state from MACI when we start processing the
+    // first batch
     if (!this.stateCopied) {
       this.copyStateFromMaci();
     }
 
     // Generate circuit inputs
     const circuitInputs = stringifyBigInts(
       this.genProcessMessagesCircuitInputsPartial(this.currentMessageBatchIndex),
-    ) as Record<string, bigint | bigint[] | bigint[][]>;
+    ) as CircuitInputs;
 
+    // we want to store the state leaves at this point in time
+    // and the path elements of the state tree
     const currentStateLeaves: StateLeaf[] = [];
-    const currentStateLeavesPathElements: any[] = [];
+    const currentStateLeavesPathElements: PathElements[] = [];
 
+    // we want to store the ballots at this point in time
+    // and the path elements of the ballot tree
     const currentBallots: Ballot[] = [];
-    const currentBallotsPathElements: any[] = [];
+    const currentBallotsPathElements: PathElements[] = [];
 
+    // we want to store the vote weights at this point in time
+    // and the path elements of the vote weight tree
     const currentVoteWeights: bigint[] = [];
-    const currentVoteWeightsPathElements: any[] = [];
+    const currentVoteWeightsPathElements: PathElements[] = [];
 
+    // loop through the batch of messages
     for (let i = 0; i < batchSize; i++) {
+      // we process the messages in reverse order
```

Comment:
Just curious about why we process messages in reverse order?

Reply:
answered on discord but if someone else is curious here are some places where to look for this answer: 

* https://maci.pse.dev/docs/key-change
* https://github.com/privacy-scaling-explorations/maci/issues/415
* https://maci.pse.dev/docs/spec#61-message-processing-circuit